libfuse/start.go

```diff
@@ -42,52 +46,54 @@ func Start(mounter Mounter, options StartOptions, kbCtx libkbfs.Context) *libfs.
 	}
 
 	log.Debug("Initializing")
-	interruptFn := func() {}
-	config, err := libkbfs.Init(
-		kbCtx, options.KbfsParams, nil, func() { interruptFn() }, log)
+	mi := libfs.NewMountInterrupter()
+	config, err := libkbfs.Init(kbCtx, options.KbfsParams, nil, mi.Done, log)
 	if err != nil {
 		return libfs.InitError(err.Error())
 	}
 	defer libkbfs.Shutdown()
 
-	log.Debug("Mounting: %s", mounter.Dir())
-	c, err := mounter.Mount()
-	if err != nil {
-		return libfs.MountError(err.Error())
-	}
-	defer mounter.Unmount()
-
-	done := make(chan struct{})
-	if c != nil { // c can be nil for NoopMounter
-		interruptFn = func() {
-			if unmountErr := mounter.Unmount(); unmountErr != nil {
-				log.Debug("Unmounting error: %v", unmountErr)
-			}
-		}
+	if options.SkipMount {
+		log.Debug("Skipping mounting filesystem")
 	} else {
-		interruptFn = func() {
-			close(done)
+		err = startMounting(config, options, log, mi)
+		if err != nil {
+			log.Errorf("Mounting filesystem failed: %v", err)
+			// Abort on error if we were force mounting, otherwise continue.
+			if options.ForceMount {
+				// If we exit we might want to clean a mount behind us.
+				mi.Done()
+				return libfs.MountError(err.Error())
+			}
 		}
 	}
+	mi.Wait()
+	return nil
+}
+
+func startMounting(config libkbfs.Config, options StartOptions,
+	log logger.Logger, mi *libfs.MountInterrupter) error {
+	log.Debug("Mounting: %q", options.MountPoint)
+	c, err := fuseMount(options)
+	if err != nil {
+		return err
+	}
+	mi.SetOnceFun(func() { fuseUnmount(options) })
```

Comment:
There could be a race condition here where `mi.Done` was called after the `fuseMount` above, but before the `SetOnceFun` here manages to grab to lock, and the mount point will never be unmounted.

Perhaps we can have a channel `doneCalled` that `mi.Done` close in a separate`Once` wrapper, before the `mi.once.Done` is called, and `SetOnceFun` would call the Fun (through `mi.once.Done`) if the channel is already closed?